File: contest/LeetCode_Weekly_Contest_204/Maximum_Length_of_Subarray_With_Positive_Product.cpp
class Solution {
public:
    int getMaxLen(vector<int>& nums) {
        int n = nums.size();
        int L = 0, R = 0, ret = 0;
        while (L < n) {
            while (R < n && nums[R] == 0) R++;
            L = R;
            int neg = 0;
            while (R < n && nums[R] != 0) {
                if (nums[R] < 0) neg++;
                if (!(neg & 1)) ret = max(ret, R - L + 1); 
                R++;
            }
            while (L < R) {
                if (!(neg & 1)) ret = max(ret, R - L);
                if (nums[L] < 0) neg--;
                L++;
            }
        }
        return ret;
    }
};

File: contest/LeetCode_Weekly_Contest_204/Maximum_Length_of_Subarray_With_Positive_Product.py
class Solution(object):
    def getMaxLen(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        L, R, n, ret = 0, 0, len(nums), 0
        
        while L < n:
            while L < n and nums[L] == 0:
                L += 1
            R = L
            
            neg = 0
            while R < n and nums[R] != 0:
                if nums[R] < 0: 
                    neg += 1
                if neg % 2 == 0:
                    ret = max(ret, R - L + 1)
                R += 1
            
            while L < R:
                if neg % 2 == 0:
                    ret = max(ret, R - L)
                if nums[L] < 0: 
                    neg -= 1
                L += 1
                
        return ret